### ISSUE TITLE
Feature request: Allow an output assembly to be generated even if there are errors in the source

### ISSUE THREAD

**Issue by erik-kallen** [dotnet/roslyn#21560]
It would be a nice feature if the compiler were able to generate an output even if there are errors. Two reasons are:

1. Some people like doing hardcore TDD, writing the test before writing the method to be tested. Currently this means that you need to create skeletons for all methods you want to test.
2. When performing large refactorings, the code often ends up in a state in which it doesn't compile for a few days. It would be much easier to do this kind of work if you could run the code during these refactorings without resorting to hacks like commenting out most of the code.

So my suggestion is that, given a compiler flag, compilation errors should not prevent the generation of an output assembly.

For example, given the method

    int M() {
        int i = 0;
        if (DateTime.Now.Year > 2017) {
            int j = ;
        }
        return 0;
    }

code could be generated as if the input were

    int M() {
        int i = 0;
        if (DateTime.Now.Year > 2017) {
            throw new CompilationErrorException("CS1525", "Invalid expression term ';'");
        }
        return 0;
    }    

Given the method

    int M() {
        int i = 0;
        if (DateTime.Now.Year > 2017) {
            int j = DateTime.Now.Bad;
        }
        return 0;
    }

code could be generated as

    int M() {
        int i = 0;
        if (DateTime.Now.Year > 2017) {
            throw new CompilationErrorException("CS1061", "'DateTime' does not contain a definition for 'Bad' and no extension method 'Bad' accepting a first argument of type 'DateTime' could be found (are you missing a using directive or an assembly reference?)");
        }
        return 0;
    }

and for a final example, Given:

    class C : IComparable {
    }

the following code could be generated:

    class C : IComparable {
        int IComparable.CompareTo(object obj) {
            throw new CompilationErrorException("CS0535", "'C' does not implement interface member 'IComparable.CompareTo(object)'");
        }
    }

**Comment by HaloFour** [comment]
Seems that the compiler would have to make a lot of assumptions regarding the remaining structure of the code in order to even approach something like this.

**Comment by scegg** [comment]
maybe this is a request about adding a source code parser layer helper object, in order to change some code into another, before passing it to syntax tree.

**Comment by erik-kallen** [comment]
@HaloFour I don't think it needs to make a lot of assumptions. I think it probably needs just three

1) When encountering a syntax error in a statement, replace that statement with a `throw` and use the standard parser recovery technique to find the next statement. We are able to get more than one syntax error in a source file, after all.
2) When members are missing (missing interface members, missing abstract overrides, possibly others), synthesize those members with bodies containing just a `throw` statement.
3) When a statement cannot be bound, replace that statement with a `throw` statement.

@scegg No, not at all. This is a suggestion to get an output assembly in the presence of errors

**Comment by scegg** [comment]
@erik-kallen 
Like tuning compile errors into runtime exception? If so, why not to code them directly or by a code transformer.

**Comment by erik-kallen** [comment]
Because I don't want to do any coding myself. What I want is "I want to run code even though the entire project does not compile".

**Comment by scegg** [comment]
@erik-kallen 
So try to build a transformer to transform the code by replacing line of error into throw. The error location can be grabbed from emit result. But a second time compile after code transformation is required.

Usually, compiler don't change any source code.

**Comment by erik-kallen** [comment]
I don't want to build a code transformer. I want the compiler to emit code that can be run unless it hits places with errors. TypeScript can do this, for example, and it is a very common advantage raised by people who advocate for dynamic languages.

**Comment by HaloFour** [comment]
@erik-kallen 

And when braces don't line up?  The entire structure of the code following a syntax error is always suspect.  Even just tossing a `throw` in there is unlikely to compile, and trying to solve the remaining issues would be a ridiculous amount of work.

C# isn't a script language.  If your program isn't structurally sound and grammatically correct it shouldn't compile.

**Comment by scegg** [comment]
@erik-kallen 
Yup, this is the problem. C#/VB.Net are not dynamic languages and MSbuild is not support this function.

**Comment by svick** [comment]
@HaloFour Roslyn attempts to fix errors like that by inserting `MissingToken`s. Though of course it doesn't always guess right, which causes further errors. But that just means tossing few more (or a lot more) `throw`s.

For example, consider this code:

```c#
class C
{
    int M()
    {
        int i = 0;
        if (DateTime.Now.Year > 2017)
        {
            int j = i;
        return 0;
    }
}
```

It's parsed as if you wrote (errors are in comments):

```c#
class C
{
    int M() // CS0161: 'C.M()': not all code paths return a value
    {
        int i = 0;
        if (DateTime.Now.Year > 2017)
        {
            int j = i;
            return 0;
        }
    }
} // error CS1513: } expected
```

The missing brace was inserted at the end of the file, instead of before the `return`, which caused the additional "missing `return`" error. But you could still generate the code like this:

```c#
class C
{
    int M()
    {
        int i = 0;
        if (DateTime.Now.Year > 2017)
        {
            int j = i;
            return 0;
        }
        throw new CompilationErrorException("CS0161", "'C.M()': not all code paths return a value");
    }
}
```

So I think it would be possible.

Whether it would be a good idea, or whether the exceptions could give reasonable information is another question. (In my example above, the exception complains about missing `return`, when the real error is missing `}`.)

**Comment by erik-kallen** [comment]
Perhaps implementing this for syntax errors wouldn't be worth it (though it would be possible; the compiler can usually find out what you mean and recover and parse the rest of the file correctly), the real value is in the other error classes

**Comment by Pilchie** [comment]
I've wanted this forever too!

Relates to https://github.com/dotnet/roslyn/issues/20914.

**Comment by jcouv** [comment]
This feature is on my radar, but I will only look at it after #20914 (tolerate errors in skeleton assemblies), which does this for APIs but not bodies. 

Besides compiler support (which by itself needs some investigation), this feature probably needs some UI in the IDE. We should also make sure that Edit&Continue allows the user to fix the code when bad code is reached.

Regarding braces: maybe we shouldn't tolerate mismatching or missing braces. It seems easy enough to give the code the correct shape before trying to run it. Allowing such badly shaped code would likely lead to bad surprises for the user anyways.

What languages support such partial compilation already? I remember some mention of a Haskell flavor, but I don't remember which.

**Comment by Pilchie** [comment]
> What languages support such partial compilation already? I remember some mention of a Haskell flavor, but I don't remember which.

VB6 did, which, combined with Edit and Continue made a great incremental development option.

**Comment by HaloFour** [comment]
@Pilchie 

> VB6 did, which, combined with Edit and Continue made a great incremental development option.

That "feature" worked spotty at best and most of the time was just a hindrance.  Much more often than not the developer accidentally left a typo in their code, something that could have otherwise been caught and resolved very quickly.  Instead the IDE would happily let the developer start their project and navigate through to whatever they had intended to test only to trip this error and break to the debugger where fixing the typo would likely exceed what E&C allowed forcing a restart of the project.  Combine that with the default behavior of not saving your project when running and the general flakiness of the VB debugger ([IPU](https://en.wikipedia.org/wiki/Invisible_Pink_Unicorn) help you if you were subclassing windows) and the likelihood of losing all of your work was exceptionally high.

Ctrl-F5 was the VB developer's best friend.  It's been nearly a decade since I last worked on classic VB professionally and I still use Ctrl-F5 in Visual Studio just out of habit/PTSD.  If you need a good reason to *not* consider this feature you don't need to look any further than VB5/6.

**Comment by Pilchie** [comment]
@HaloFour interesting - that certainly doesn't match my (or most other VB6 developers that I talk to) experience, but it's a great datapoint to have.  I don't think we want to stop presenting the error in the error-list or anything like that - just all you to make forward progress (and build other, dependent projects), etc.

**Comment by HaloFour** [comment]
@Pilchie 

We're going to have to agree to disagree on that.  There are numerous things that I think VB5/6 did well, the debugging experience was not one of them.  It didn't help that the debugger was an in-proc interpreter where it wasn't all that difficult to cause the program and the IDE to crash.  Anyone who has ever subclassed a window from VB5/6 can testify to this fact because if you ever tripped the syntax error message box (because you didn't compile with Ctrl-F5) the IDE would immediately and catastrophically crash.  All of that code you wrote within E&C was totally lost, as could all of the code you wrote in that session assuming that you left the default options in place and you didn't explicitly save.

But much more importantly, for all of it's foibles as an IDE and as much as I disagree with using a debugger in this manner, it was never possible in VB5/6 to actually produce a binary that wouldn't compile.  This entire experience was driven strictly by the interpreter.  That is a big difference from what this proposal is suggesting, where the compiler will have to spit out a perfectly fine looking assembly that happens to be riddled with compile-time errors that won't surface until run time.  Sure, these compile-time errors would be displayed in the Error List, but that thing is loaded with tons of noise in any reasonably large solution.  I've got a single project solution right now that shows errors in the Error List but still builds just fine.  At the end of the day the developer will see "Build Succeeded" and think that they're done.  That very possibility should scare any reasonable developer to death.

If you want to emulate the better points of VB5/6 in Visual Studio I'd suggest working on getting that load time sub-second.  Visual Studio 2017 cold loads in about 20 seconds on my desktop.  Creating a new empty project is another 5 or so.  And the IDE still doesn't handle unsaved projects well, e.g. the NuGet Package Manager can't work with them at all.

**Comment by erik-kallen** [comment]
It probably would be a good idea to make this feature opt-in, which in the normal case would be controlled by an msbuild variable.

That way you can add `/p:EmitOutputAssemblyInCaseOfCompileErrors=false` to the msbuild invocation on your build server, and your build will never succeed regardless of if people enable it in their projects.

But if your project shows build errors without anyone bothering to fix them, I suggest you start looking at your development methodology.
